staticaddr/loopin/actions.go

```diff
@@ -0,0 +1,1053 @@
+	// only be the case where a user-provided payment timeout was hit, the
+	// invoice got canceled and the timeout of the htlc was not reached yet.
+	// So we want to wait until the htlc timeout path opens up so that we
+	// could sweep the funds back to us if the server published it without
+	// paying the invoice.
+	subscribeCtx, cancelInvoiceSubscription := context.WithCancel(ctx)
+	defer cancelInvoiceSubscription()
+
+	invoiceUpdateChan, invoiceErrChan, err :=
+		f.cfg.InvoicesClient.SubscribeSingleInvoice(
+			subscribeCtx, f.loopIn.SwapHash,
+		)
+
+	if err != nil {
+		err = fmt.Errorf("unable to subscribe to swap "+
+			"invoice: %w", err)
+
+		return f.HandleError(err)
+	}
+
+	htlc, err := f.loopIn.getHtlc(f.cfg.ChainParams)
+	if err != nil {
+		err = fmt.Errorf("unable to get htlc: %w", err)
+
+		return f.HandleError(err)
+	}
+
+	// Subscribe to htlc tx confirmation.
+	reorgChan := make(chan struct{}, 1)
+	registerHtlcConf := func() (chan *chainntnfs.TxConfirmation, chan error,
+		error) {
+
+		return f.cfg.ChainNotifier.RegisterConfirmationsNtfn(
+			ctx, nil, htlc.PkScript, defaultConfTarget,
+			int32(f.loopIn.InitiationHeight),
+			lndclient.WithReOrgChan(reorgChan),
+		)
+	}
+	htlcConfChan, htlcErrConfChan, err := registerHtlcConf()
+
+	if err != nil {
+		err = fmt.Errorf("unable to monitor htlc tx confirmation: %w",
+			err)
+
+		return f.HandleError(err)
+	}
+
+	// Subscribe to new blocks.
+	registerBlocks := f.cfg.ChainNotifier.RegisterBlockEpochNtfn
+	blockChan, blockChanErr, err := registerBlocks(ctx)
+	if err != nil {
+		err = fmt.Errorf("unable to subscribe to new blocks: %w", err)
+
+		return f.HandleError(err)
+	}
+
+	htlcConfirmed := false
+
+	// Create the swap payment timeout timer. If it runs out we cancel the
+	// invoice, but keep monitoring the htlc confirmation.
+	// If the invoice was canceled, e.g. before a restart, we don't need to
+	// set a new deadline.
+	invoice, err := f.cfg.LndClient.LookupInvoice(ctx, f.loopIn.SwapHash)
+	if err != nil {
+		err = fmt.Errorf("unable to look up invoice by swap hash: %w",
+			err)
+
+		return f.HandleError(err)
+	}
+
+	var serverPaymentDeadline time.Timer
+	if invoice.State != invoices.ContractCanceled {
+		// If the invoice is still live we set the timeout to the
+		// remaining payment time. If too much time has elapsed, e.g.
+		// after a restart, we set the timeout to 0 to cancel the
+		// invoice and unlock the deposits immediately.
+		remainingTimeSeconds := f.loopIn.RemainingPaymentTimeSeconds()
+		var timeout time.Duration
+		if remainingTimeSeconds > 0 {
+			timeout = time.Duration(remainingTimeSeconds) *
+				time.Second
+		} else {
+			// If the invoice isn't cancelled yet and the payment
+			// timeout elapsed, we set the timeout to 0 to cancel
+			// the invoice and unlock the deposits immediately.
+			timeout = time.Duration(0) * time.Second
+		}
+		serverPaymentDeadline = *time.NewTimer(timeout)
```

Comment:
Can `serverPaymentDeadline` be of type `time.Timer`? This dereference would not be needed. (I'm always suspicious about such dereferences of types whose `New` funcion return pointers. Some of them don't support copying.)

Reply:
As discussed, if `serverPaymentDeadline` is not set it results in a nil pointer in the select(`serverPaymentDeadline.C`). As suggested will create `deadlineChan := serverPaymentDeadline.C` and select on that to avoid the nil pointer.